### PR TITLE
Insert row

### LINKED ISSUE #9269: Implement HTMLTableElement.insertRow()
Note that the WebIDL definition for the `HTMLTableElement` has changed, so an edit of the WebIDL file is necessary.

Code: `components/script/dom/htmltableelement.rs`, `components/script/dom/webidls/HTMLTableElement.webidl`
Tests: `tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-01.html`,  `tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-02.html`
Spec: https://html.spec.whatwg.org/multipage/#dom-table-insertrow, https://html.spec.whatwg.org/multipage/#htmltableelement

Blocks #2968.

### PR DESCRIPTION
**Pull request by g-k**
Fixes #9269

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10508)

### CODE REVIEW

**Comment by KiChjang** [comment]
Great job! I'm not entirely sure we need all those `.expect()` there though... @nox or @jdm, do you think an `.unwrap()` is good enough for these use cases? I'm mostly concerned about polluting the error logs.

---

Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmltableelement.rs, line 143 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz58VRyzAm5KfQESc0:-KEz58VRyzAm5KfQESc1:1285204257) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L143)):*
`Fallible<Root<HTMLTableRowElement>>`

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KEz7M4kGhbY8xCt_Ymm:1347988085) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
This doesn't look correct. The new_row should be inserted before the _i_th element, not appended as a child to its parent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508#-:-KEz6qMx0GYiFfH-9Wu4:-841201697)*

**Comment by jdm** [comment]
I have no problem with choosing expect over unwrap.

**Comment by frewsxcv** [comment]
> Great job! I'm not entirely sure we need all those .expect() there though...

All the `expect()`s on `AppendChild`s should probably use `try!`, right?

**Comment by g-k** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmltableelement.rs, line 143 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz58VRyzAm5KfQESc0:-KF-tsYhfEWdRKBgZGGB:-499146009) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L143)):*
Done.

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KF-tzeRY48JSM8a18kS:-1103506662) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
d'oh! Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508)*

**Comment by g-k** [comment]
Let me know if I should update the Option unwrapping to be more concise with `try!` or `unwrap`.

**Comment by KiChjang** [comment]
@g-k I think it's better to use `try!` here, since the method is explicitly marked as `Fallible`

**Comment by KiChjang** [comment]
-S-awaiting-review  +S-needs-code-changes 

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/domexception.rs, line 105 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10508#-KF2TquGpZ4FpAVb4DQG:-KF2TquGpZ4FpAVb4DQH:-119683211) ([raw file](https://github.com/servo/servo/blob/df7c572dd595b01a31d698c1cb19947791ff81da/components/script/dom/domexception.rs#L105)):*
Is there a spec link that defines this `UnknownError` type?

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KF2TIKuYivrfOzMNcj8:339235214) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
This is still not fixed; you're inserting the new node before the parent, not before the _i_th row.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508#-:-KF2U83WUIugv8bAVmvc:-1312997665)*

**Comment by Ms2ger** [comment]
I'm sorry, but @KiChjang is wrong in suggesting `try!()` here. If the logic of the function ensures that the operation will always be successful, we should `.expect()` or `.unwrap()`.

**Comment by notriddle** [comment]
You should probably run `git rebase -i --autosquash upstream/master && git push -f`. Put your fixup commits into their base commits.

**Comment by frewsxcv** [comment]
Just going to point out that @g-k force pushed commits but they don't look squashed

**Comment by KiChjang** [comment]
Reviewed 1 of 2 files at r5, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/webidls/HTMLTableElement.webidl, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSUqtHjSXTSVeEPMH:-KFLSUquM67BHy7K0FPC:643673610) ([raw file](https://github.com/servo/servo/blob/ce805f7497a97c22c1e955222453d76d4d055222/components/script/dom/webidls/HTMLTableElement.webidl#L9)):*
Do you mind going to the spec WebIDL and copying the current definitions over?

---

*[tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-03.html, line 3 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSgQ7tZpCLhjtrFc2:-KFLSgQ7tZpCLhjtrFc3:-1948119440) ([raw file](https://github.com/servo/servo/blob/cfda50d62c5aa032a49e32ed1ae1ccbafe83a6c6/tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-03.html#L3)):*
That doesn't look like it's your email.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508)*

**Comment by g-k** [comment]
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KF2nXSjOnDL5ph998vt:718172292) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
The wpt test I added passes, but I didn't test the new test against other browser engines.

---

*[components/script/dom/webidls/HTMLTableElement.webidl, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSUqtHjSXTSVeEPMH:-KFLgnjgoTB86lflWkZ5:1486914206) ([raw file](https://github.com/servo/servo/blob/ce805f7497a97c22c1e955222453d76d4d055222/components/script/dom/webidls/HTMLTableElement.webidl#L9)):*
I don't think anything else changed: https://html.spec.whatwg.org/multipage/tables.html#htmltableelement

---

*[tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-03.html, line 3 \[r8\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSgQ7tZpCLhjtrFc2:-KFLhAgeYU3SBHmYpdeh:-984926693) ([raw file](https://github.com/servo/servo/blob/cfda50d62c5aa032a49e32ed1ae1ccbafe83a6c6/tests/wpt/web-platform-tests/html/semantics/tabular-data/the-table-element/insertRow-method-03.html#L3)):*
ah stupid copy paste mistake

---

*[components/script/dom/domexception.rs, line 105 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10508#-KF2TquGpZ4FpAVb4DQG:-KF2dZSsq2PqbCB5pdzc:-1711191681) ([raw file](https://github.com/servo/servo/blob/df7c572dd595b01a31d698c1cb19947791ff81da/components/script/dom/domexception.rs#L105)):*
yep, https://heycam.github.io/webidl/#unknownerror

Not sure if the link should be in a comment, but I put it in the commit message: https://github.com/servo/servo/pull/10508/commits/a15b814bae1d6cc6d35c94a3c2cd08ac909eb70e

I'm also not sure if another error type would make more sense. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508)*

**Comment by KiChjang** [comment]
Reviewed 1 of 3 files at r6, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KFLiOpuf8u6RqNx-UaF:1773586402) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
This is fine now, I misread the usage of InsertBefore.

---

*[components/script/dom/webidls/HTMLTableElement.webidl, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSUqtHjSXTSVeEPMH:-KFLiThtGjc0Z9mwhiGd:-1487456523) ([raw file](https://github.com/servo/servo/blob/ce805f7497a97c22c1e955222453d76d4d055222/components/script/dom/webidls/HTMLTableElement.webidl#L9)):*
This createCaption method should return an `HTMLTableCaptionElement` now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508)*

**Comment by g-k** [comment]
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmltableelement.rs, line 190 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10508#-KEz7M4kGhbY8xCt_Yml:-KFLl3Fgw9gtv-RmxHD8:-510466975) ([raw file](https://github.com/servo/servo/blob/9f33c6d2405b8b4935ead0c614ae2a74e565555c/components/script/dom/htmltableelement.rs#L190)):*
:+1: 

---

*[components/script/dom/webidls/HTMLTableElement.webidl, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10508#-KFLSUqtHjSXTSVeEPMH:-KFLl5XqKOrrO7LyuMce:-1138445948) ([raw file](https://github.com/servo/servo/blob/ce805f7497a97c22c1e955222453d76d4d055222/components/script/dom/webidls/HTMLTableElement.webidl#L9)):*
oh yeah, good eye, I missed that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10508)*

**Comment by KiChjang** [comment]
Still needs squash.

**Comment by KiChjang** [comment]
fn CreateCaption needs to be updated to return a Root<HTMLTableCaptionElement>.

File: components/script/dom/htmltableelement.rs [hunk 1]
```diff
@@ -177,7 +178,7 @@ impl HTMLTableElementMethods for HTMLTableElement {
                 caption
             }
         };
-        Root::upcast(caption)
+        caption
```
Comment (KiChjang) [inline components/script/dom/htmltableelement.rs]: nit: This line can be removed altogether, and the let-binding removed, making this function contain only a match expression instead.

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!

**Comment by KiChjang** [comment]
@bors-servo r-

Sorry, this still needs to be squashed. Your first commit alone will cause build errors.

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!

**Comment by KiChjang** [comment]
@bors-servo r-

The test expectations need to be updated.

**Pull request merged**